graylog2-server/src/main/java/org/graylog/plugins/pipelineprocessor/processors/PipelineInterpreter.java

```diff
@@ -93,6 +97,9 @@ public PipelineInterpreter(Journal journal,
     public Messages process(Messages messages) {
         try (Timer.Context ignored = executionTime.time()) {
             final State latestState = stateUpdater.getLatestState();
+            if (latestState.enableRuleMetrics()) {
+                return process(messages, new RuleMetricsListener(metricRegistry), latestState);
```

Comment:
Using a `LocalMetricRegistry` doesn't help in this case. We cannot reliably remove the timer metrics once the user disabled the metric collection because we don't have a synchronization point where all processing threads use the updated state.

That means even if we remove the timer metrics after disabling the collection, we don't know if one thread is still using the old state and re-creates the metrics again.

The metrics are cleaned up when a rule gets deleted or the server gets restarted.